### DISCUSSION TITLE
Issue with Custom MPS Implementation

### DISCUSSION THREAD

**Discussion by rezah** [netket/netket#2071] [General]
I am unable to obtain converged results using the simple custom MPS defined below. The variational energy fails to approach the exact result, even when using a large bond dimension. I suspect I may be making a naïve mistake in the implementation. Notice that contration should be right

```
import netket as nk
from flax import nnx
import jax
import jax.numpy as jnp
from netket.operator.spin import sigmax, sigmaz

class vmps_(nnx.Module):
    def __init__(self, *, L: int, bond_dim: int, rngs: nnx.Rngs):
        self.L = L
        self.bond_dim = bond_dim
        D = bond_dim
        self.tensors: List[nnx.Param] = []
        for i in range(L):
            Dl = 1 if i == 0 else D
            Dr = 1 if i == L - 1 else D
            shape = (2, Dl, Dr)
            key = rngs.params()
            tensor = 0.01 * jax.random.normal(key, shape, dtype=jnp.complex128)
            self.tensors.append(nnx.Param(tensor))
    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:

        def single_log_amp(x_sample):
            A0 = self.tensors[0]
            psi = A0[x_sample[0]]  # shape (D_right,)

            for i in range(1, self.L):
                Ai = self.tensors[i]
                psi = jnp.tensordot(psi, Ai[x_sample[i]], axes=[-1, -2])  # contract right bond

            amp = psi.squeeze()  # should be scalar
            return jnp.log(amp + 1e-12 + 0j)  # safe complex log

        return jax.vmap(single_log_amp)(x)

L=4
model = vmps_(L=L, bond_dim=4, rngs=nnx.Rngs(1))
hilbert = nk.hilbert.Spin(s=1/2, N=L)

Gamma = -1
H = sum([Gamma * sigmax(hilbert, i) for i in range(L)])
V = -1
H += sum([V * sigmaz(hilbert, i) * sigmaz(hilbert, (i + 1) % L) for i in range(L)])

sampler = nk.sampler.MetropolisLocal(hilbert=hilbert)
vstate = nk.vqs.MCState(sampler, model, n_samples=10000)

gs = nk.VMC(H, nk.optimizer.Adam(learning_rate=0.01), variational_state=vstate)
gs.run(100)
```

<img width="549" alt="image" src="https://github.com/user-attachments/assets/56687623-006b-46de-91e4-a2c5512d9061" />

**Comment by PhilipVinc** [comment]
can you share a full script that runs?

**Reply by PhilipVinc** [reply]
The issue is that your gradient at initialisation is 0. Maybe you are at an eigenstate?

```python
In [15]: e,g=vstate.expect_and_grad(H)

In [16]: jax.tree.map(jnp.linalg.norm, g)
Out[16]:
{'tensors': {0: Array(1.92415167e-13, dtype=float64),
  1: Array(2.14330389e-13, dtype=float64),
  2: Array(1.1715484e-13, dtype=float64),
  3: Array(8.52686205e-14, dtype=float64)}}
```

**Reply by PhilipVinc** [reply]
Your state is uniform

```python
In [41]: vstate.to_array()
Out[41]:
array([-0.17517561+0.17836341j, -0.17517561+0.17836341j,
       -0.17517561+0.17836341j, -0.17517561+0.17836341j,
       -0.17517561+0.17836341j, -0.17517561+0.17836341j,
       -0.17517561+0.17836341j, -0.17517561+0.17836341j,
       -0.17517561+0.17836341j, -0.17517561+0.17836341j,
       -0.17517561+0.17836341j, -0.17517561+0.17836341j,
       -0.17517561+0.17836341j, -0.17517561+0.17836341j,
       -0.17517561+0.17836341j, -0.17517561+0.17836341j])
```
which is a bit weird.
I suspect your implementation is wrong.

Have a look at `nk.models.tensor_networks.MPSPeriodic` for some inspiration maybe?

**Reply by PhilipVinc** [reply]
@rezah the bug is the following: in your code you index with the sample
`Ai[x_sample[i]]`

so the samples should be either 0 or 1. However
```python
In [63]: hilbert.local_states
Out[63]: [1, -1]
#or
In [64]: hilbert.all_states()
Out[64]:
Array([[ 1,  1,  1,  1],
       [ 1,  1,  1, -1],
       [ 1,  1, -1,  1],
```
the states of the Spin Hilbert space are encoded as +/-1 which then breaks your code which assumes 0 or 1.

To fix it you can either use the function `hilbert.states_to_local_indices` which converts your local values in the Hilbert space to integer 0...D values, 
```python
    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        x = hilbert.states_to_local_indices(x)
        def single_log_amp(x_sample):
            A0 = self.tensors[0]
            psi = A0[x_sample[0]]  # shape (D_right,)
```
or just use a `nk.vqs.Qubit(L)` Hilbert space which encodes stuff as 0 and 1

**Reply by rezah** [reply]
Great catch, thank you @PhilipVinc!
